Initial support for DateTime types

eminence opened the pull request:
This is an initial draft at adding DateTime handling into numbat.  This is incomplete, but I'm opening the PR now to start a discussion about the general approach.

Some review notes:

* DateTimes didn't really feel like any of the existing types, so they are an entirely new thing (so the parser, tokenize, AST, etc all are updated accordingly)
* At the moment, the only way to construct a DateTime object is with the new `now()` function, but I envision some new date parsing functions (like a `strptime`) at a minimum, and maybe dedicated handling in the parser for well-formatted dates
* Two operations on DateTimes have been implemented: 
  *  Adding/Subtracting a duration (something of the `Time` dimension) to a `DateTime` to produce another `DateTime`
  * Subtracting two `DateTime`s to produce something of type `Time`.

  Because of this dependency on having a `Time` dimension, the typechecker has some special code to check for times, which feels a bit ugly (but might be unavoidable)

  What other operations can we perform on dates?
* Controlflow in typechecker is a bit of a mess as written, and deserves to be improved somehow
* The `BinaryOperatorForDate` variant needs to be redesigned, I think.
* Tests and error handling is not added yet
* I'm using `chrono` as the underlying date/time library.  `time` is another good choice.  I picked `chrono` just because I have more personal experience with it.  Both crates support wasm as far as I know.

Small demo:

```
>>> let x = now()

  let x: DateTime = now()

>>> x

  x

    = 2023-12-31T17:09:48.505882820Z    [DateTime]

>>> now() - x -> milliseconds

  now() - x ➞ millisecond

    = 9088.02 ms    [Time]

>>> x + 2 days + 4 seconds

  (x + 2 day) + 4 second

    = 2024-01-02T17:09:52.505882820Z    [DateTime]

>>> now() + (180 miles / 55 mph)

  now() + (180 mile / 55 mph)

    = 2023-12-31T20:26:54.595755947Z    [DateTime]

>>>
```

CC #181

sharkdp commented:
This is just awesome — thank you very much for looking into this!



> * DateTimes didn't really feel like any of the existing types, so they are an entirely new thing (so the parser, tokenize, AST, etc all are updated accordingly)

Exactly what I envisioned :+1: 

> some new date parsing functions (like a `strptime`) at a minimum

:+1: 

> and maybe dedicated handling in the parser for well-formatted dates

I think this could be a second step maybe (and potentially needs some thought), to keep this PR manageable?



> Two operations on DateTimes have been implemented:
> 
>     * Adding/Subtracting a duration (something of the `Time` dimension) to a `DateTime` to produce another `DateTime`
> 
>     * Subtracting two `DateTime`s to produce something of type `Time`.
> 
> 
> Because of this dependency on having a `Time` dimension, the typechecker has some special code to check for times, which feels a bit ugly (but might be unavoidable)

Ah, right. That is a bit unfortunate, yes. So far, we managed to decouple the unit- and dimension system (the prelude) fairly well from the language implementation. Maybe we can still achieve this if we 'register' the `Time` dimension somehow (like with a `@`-decorator when defining `dimension Time`)? Or maybe it helps to introduce a hidden `__Time` dimension (I currently don't see how :smile:)? 

I also have a somewhat related question: are the current quantities of dimension `Time` really suitable as durations? A `year`, for example, is defined as an average (Gregorian) year with a length of `365.2425 days`. So running `now() +  1 year` *today* (Jan 2nd, 2024... not too late) will actually show me the Jan 1st, 2025, approximately 6 hours later than it is now. Because (a) 2024 is a leap year, i.e. has 366 days and (b) those `0.2425 days` will shift the time by 5.82 hours.

A similar problem appears when we add/subtract `months`. But most users would probably expect the datetime to be shifted by a year *from now* (i.e. the exact same datetime, but with 2025 instead of 2024) or by a month *from now*. So we probably need to distinguish between years and calendar years. And similar for months.

I'm bringing this up here because we might need *another* special type anyway (maybe `Duration` vs `Time`).

A much more conservative approach would be to only have functions like `add_months(dt: DateTime, num_months: Scalar) -> DateTime`. But this would also render this much less useful.

What do you think?

> What other operations can we perform on dates?

Despite what I said above, I think what you have already is really powerful! Calculations like `now() + 300 minutes` or `now() + 7.5 hours` come up all the time and already work as expected. And the other big thing (for me) are differences, which also work... and would be *really* useful if we had a special syntax in the parser like `@17:15 - @8:40` (*"how many hours have I worked today?"*) or @2023-12-14 - @2023-07-25 (*"how many days are there between Dec 14th and July 25th?"*).

The other thing I do frequently is to convert from/to UNIX timestamps, but that should be fairly easy to support with FFI functions.

And then there are timezones. I would certainly like to see datetimes in the local timezone by default. But it would be great if we could also convert to other timezones somehow.

> * I'm using `chrono` as the underlying date/time library.  `time` is another good choice.  I picked `chrono` just because I have more personal experience with it.  Both crates support wasm as far as I know.

:+1:

sharkdp commented on numbat/modules/core/datetime.nbt:
```diff
@@ -0,0 +1 @@
+fn now() -> DateTime
```
If we do `numbat --no-prelude`, and then only import `use core::datetime`, we don't have a `Time` dimension. So doing something like
```
use core::datetime
now() - now()
```
will crash.

Could be resolved for now by importing `core::dimensions`?


Minor nitpick: add a trailing newline?

eminence replied:
Sure, I'll try that.  Also, I intend to have some type of nice error message like "a Time dimension was not found" as a hint for people who are building their own custom preludes.

eminence replied:
Unfortunately importing `core::dimensions` is not sufficient, because `Seconds` are needed, so I'm importing `units::si`.  It feels wrong for something in `core` to import something in `units`, though...

sharkdp commented on numbat/src/typechecker.rs:
```diff
@@ -487,134 +517,182 @@ impl TypeChecker {
                     } else {
-                        Ok(Type::Dimension(lhs_type))
+                        return Err(TypeCheckError::IncompatibleType(IncompatibleTypeError {
+                            expected_name: "a unit of Time",
```
This usually contains a dimension name, not a unit. This error message is also currently shown when doing something stupid like `now() -> seconds`, which is confusing.

eminence replied:
Yeah, the error handling and error formatting needs some attention.  The new `IncompatibleTypeError` type started as a copy/paste from `IncompatibleDimensionsError` and so it contains some vestigial fields that might not be needed, or might not be named/used correctly

sharkdp commented on numbat/src/typechecker.rs:
```diff
@@ -487,134 +517,182 @@ impl TypeChecker {
+                            Type::Dimension(time),
+                            true,
+                        )
+                    } else if *op == BinaryOperator::Add
+                        || *op == BinaryOperator::Sub && rhs_is_time
```
This should probably be
```suggestion
                    } else if (*op == BinaryOperator::Add
                        || *op == BinaryOperator::Sub) && rhs_is_time
```
because addition is currently not properly type-checked (try `now() + 1 meter`).

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -40,13 +40,33 @@ impl DType {
     }
+    /// Is the current dimension type the Time dimension?
+    ///
+    /// This ia special helper that's useful when dealing with DateTimes
```
Minor: typo

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -40,13 +40,33 @@ impl DType {
+    ///
+    /// This ia special helper that's useful when dealing with DateTimes
+    pub fn is_time_dimension(&self, registry: &DimensionRegistry) -> bool {
+        let mut names = vec![];
+
+        let factors: Vec<_> = self.iter().collect();
+        if factors.len() == 1 && factors[0].1 == Exponent::from_integer(1) {
+            names.push(factors[0].0.clone());
+        }
+
+        names.extend(registry.get_derived_entry_names_for(self));
+
+        match &names[..] {
+            [] => false,
+            [single] if single == "Time" => true,
+            _ => false,
+        }
```
It's probably for a reason, but this seems a bit convoluted/complicated. I would have thought something like `self == BaseRepresentation::from_factor(BaseRepresentationFactor("Time".into(), Rational:::from_integer(1)))` might work here?

eminence replied:
No good reason for this.  I just copy/pasted this code from `to_readable_type()` function, without fully understanding it.  I'll try your suggestion

eminence commented:
Thanks for the initial review!

> > and maybe dedicated handling in the parser for well-formatted dates

> I think this could be a second step maybe (and potentially needs some thought), to keep this PR manageable?

Agreed.

> Maybe we can still achieve this if we 'register' the Time dimension somehow (like with a @-decorator when defining dimension Time)?

I do think something like this is needed, eventually (but maybe for a follow-up PR).  Currently, if someone is building their own prelude that has a time-like dimension that isn't called "Time" (or has a different base-unit than Second), DateTimes won't work properly.   But I'd like this to be possible.


> I also have a somewhat related question: are the current quantities of dimension `Time` really suitable as durations?

Great question.   In your `now() + 1 year` example, I agree that adding 365.243 days is probably unexpected.   As another example, consider what a user might intend with `now() + 1 month` ?  Probably most people mean something like `now.month += 1` with some unspecific behavior about what happens when the next month doesn't have the same number of days as the current month.

BTW, there is an argument here that the current prelude definitions for `year` and `month` are not obvious to most people, and maybe should be renamed (`gregorian_year`?)

I will say, however, that the current behavior of `now() + 1 year` being identical to `now() + 365.2425 days` is entirely expected and consistent given the definition of `1 year`.  So if we wanted to keep the current behavior, I think that is defensible.   In fact, I think this could be combined with your `add_months(dt: DateTime, num_months: Scalar) -> DateTime` idea.  The documentation for `add_months` could be a good place to describe the nuances here and why `now() + 1 month` behaves as it does and how if you want something else, then use `add_months`

> would be really useful if we had a special syntax in the parser like `@17:15 - @8:40` (*"how many hours have I worked today?"*) or `@2023-12-14 - @2023-07-25` (*"how many days are there between Dec 14th and July 25th?"*).

I think these two examples suggest that we also need `Date` and `Time` objects (in addition to `DateTime`).  Does that sound right?


> And then there are timezones. I would certainly like to see datetimes in the local timezone by default. But it would be great if we could also convert to other timezones somehow.

For this, I'm wondering if we can re-purpose the conversion operator?  For example, something like `now() -> GMT-0500` ?

ion1 commented:
The same concern also applies to days: either it's always equal to 24 hours or you take DST and leap seconds into account.

eminence commented:
I added a few more commits that add some new datetime functions:
* `parse_datetime`
* `format_datetime`
* `to_unixtime`
* `from_unixtime`

I also added experimental timezone support using the `->` conversion operator.  It uses strings as the right-hand-side because that allowed for easy implementation/experimentation, but I'm not necessarily sold on this syntax.  Here's what it looks like:

```
>>> now()

  now()

    = Sat, 6 Jan 2024 23:29:25 -0500    [DateTime]

>>> now() -> "CET"

  now() ➞ "CET"

    = Sun, 7 Jan 2024 05:29:34 +0100    [DateTime]

>>> now() -> "Asia/Tokyo"

  now() ➞ "Asia/Tokyo"

    = Sun, 7 Jan 2024 13:29:45 +0900    [DateTime]

>>> parse_datetime("Wed, 20 Jul 2022 21:52:05 +0200") -> "local"

  parse_datetime("Wed, 20 Jul 2022 21:52:05 +0200") ➞ "local"

    = Wed, 20 Jul 2022 14:52:05 -0500    [DateTime]
```

eminence commented:
I've been continuing to test drive this over the past week, and continue to find it very useful.

* I've simplified the error handling on typecheck errors.  Doing something like `now() + 4 meters` results in a generic `Operator  +  can not be applied to these types` error.  Unfortunately `4 meters + now()` results in a different error message, but I'm hoping that can be fixed later and won't block this PR
* So far, 100% of my uses of adding durations to dates have been with short durations (measured in days or hours), so I've not personally run into any annoyances with the strange "month" or "year" units.
  * Doing things like `now() - 8 hours - 42 minutes` feels fine.  If syntax like `now() - 8h42m` existed I would probably use it, but I don't feel like it's needed for this PR
 * Using the `->` conversion operation (with strings to represent timezones) still feels like a bit of an abuse of this operator, it's starting to feel pretty natural after a week of use (and it's exceedingly useful)
 * I tested in wasm and everything seemed to work.

At this point, I think this PR is ready for another review and wider testing.

sharkdp commented:
Thank you very much for the updates. I currently don't have the energy to review this PR, but will hopefully do so soon. Concerning the question with calendar time units vs 'averaged' time units, I think it would be cool if someone could do some investigation how other tools handle this (e.g. Qalculate, Frink, GNU units, …)

eminence commented:
Thank you for the update.  Please do not feel any rush or pressure to review this.

I've opened a new discussion thread in #292 to track the research task into how other tools handle time units

sharkdp commented on numbat/Cargo.toml:
```diff
@@ -29,6 +29,8 @@ libc = "0.2.152"
 rust-embed = { version = "8.2.0", features = ["interpolate-folder-path"] }
 num-format = "0.4.4"
 walkdir = "2"
+chrono = "0.4.31"
+chrono-tz = "0.8.5"
```
Unfortunately, this has a quite drastic effect on the WASM size.

Before: 872.45 KB (872,448 bytes)
After: 1.89 MB (1,892,352 bytes)

I'm not saying this is a no-go, but maybe we could look into reducing the size of those dependencies  (via feature flags?).

eminence replied:
Unfortunately I wasn't able to reduce the WASM size, even when disabled default features.  chrono-tz has a way to [limit the size](https://github.com/chronotope/chrono-tz?tab=readme-ov-file#limiting-the-timezone-table-to-zones-of-interest) of the TZ table which we might consider, but I'm not quite sure what ones we might omit (and I'm not currently sure how much space that would save anyway, compared to the size of chrono).   It also recommends using LTO, but we're already using that in our release profile.

sharkdp commented on numbat/modules/core/datetime.nbt:
```diff
@@ -0,0 +1,7 @@
+use units::si
+
+fn now() -> DateTime
+fn parse_datetime(input: String) -> DateTime
```
It might be useful to add
```
fn parse_date(date: String) -> DateTime = parse_datetime("{date}T00:00:00Z")
```
here (or similar), so we can do things like `parse_date("2024-02-06")`.

sharkdp replied:
Just a note: We can even implement `parse_time` in Numbat with hacks like:
```py
fn today() -> String = format_datetime("%Y-%m-%d", now())
fn parse_time(time: String) -> DateTime = parse_datetime("{today()}T{time}Z")

parse_time("12:30:00")
```

sharkdp commented on numbat/src/ffi.rs:
```diff
@@ -740,3 +785,60 @@ fn chr(args: &[Value]) -> Result<Value> {
+    let input = args[0].unsafe_as_string();
+
+    // Try to parse as rfc3339 and if that fails then as rfc2822
+    let output = chrono::DateTime::parse_from_rfc3339(input)
+        .or_else(|_| chrono::DateTime::parse_from_rfc2822(input))
```
Not for now, but it would be cool to allow other formats here as well.

sharkdp commented on numbat/src/ffi.rs:
```diff
@@ -740,3 +785,60 @@ fn chr(args: &[Value]) -> Result<Value> {
+fn format_datetime(args: &[Value]) -> Result<Value> {
+    assert!(args.len() == 2);
+
+    // TODO should we support taking the format/datetime args in either order?
```
No, I don't think so.

sharkdp commented on numbat/src/vm.rs:
```diff
@@ -615,6 +652,59 @@ impl Vm {
+                    let rhs = self.pop_string();
+                    let lhs = self.pop_datetime();
+
+                    // TODO how to handle errors, and is this `chrono_tz` crate the best choice?
```
We should at least make a wrong timezone a runtime error here, if not possible earlier. Right now, if you do `now() -> "Florida"`, it will fall back to UTC silently.

sharkdp commented on numbat-cli/src/completer.rs:
```diff
@@ -73,6 +74,40 @@ impl Completer for NumbatCompleter {
             ));
         }
 
+        // does it look like we're tab-completing a timezone (via the conversion operator)?
```
Having tab completion support for time zones is really great. I was wondering if there would be harm in doing this for all strings, not just string after `->`.

Time zones might end up in other places. And we also have `to`, `➞`, … as conversion operators.

eminence replied:
I've extended this code to recognize the other conversion operators.  I guess we could tab-complete timezones in all strings, but I think I would defer that change until timezones can actually appear in other places

sharkdp commented on numbat/src/bytecode_interpreter.rs:
```diff
@@ -112,6 +116,28 @@ impl BytecodeInterpreter {
+                        BinaryOperator::Add => Op::AddDateTime,
+                        BinaryOperator::Sub => Op::SubDateTime,
+                        BinaryOperator::ConvertTo => Op::ConvertDateTime,
+                        _ => panic!("{operator:?} is not valid with a DateTime"), // should be unreachable, because the typechecker will error first
```
Ok. Let's use `unreachable!` in that case.

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -122,10 +135,23 @@ pub enum Expression {
+        /// RHS may be a DateTime, a Time unit, or a String
+        Box<Expression>,
+        Type,
+        /// If true, then this is a date subtration operation
```
minor: typo

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -122,10 +135,23 @@ pub enum Expression {
+        Box<Expression>,
+        Type,
+        /// If true, then this is a date subtration operation
+        bool,
```
I'd be in favor of an enum argument inside the AST. Even if it only has two options.

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -122,10 +135,23 @@ pub enum Expression {
     Boolean(Span, bool),
     Condition(Span, Box<Expression>, Box<Expression>, Box<Expression>),
     String(Span, Vec<StringPart>),
+    DateTime(Span, chrono::DateTime<chrono::Utc>),
```
What does the hardcoded `chrono::Utc` mean here?

eminence replied:
This means that the `DateTime` object is always storing datetimes with the UTC offset.    When converted into a `Value::DateTime` to be displayed, it stored with the same `DateTime<Utc>` but also includes an offset to be used when pretty printing (see value.rs line 9)

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -567,6 +606,7 @@ impl PrettyPrint for Expression {
                     + m::space()
                     + with_parens(else_)
             }
+            DateTime(_, _) => m::text("datePlaceHolder"),
```
to do

eminence replied:
I actually wasn't able to figure out how to get into this code, so I'm now just printing the debug representation of the `DateTime` object.  Does that sound OK?

eminence replied:
Fully removed in the latest commits

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -122,10 +142,23 @@ pub enum Expression {
+    BinaryOperatorForDate(
+        Option<Span>,
+        BinaryOperator,
+        /// LHS must be a DateTime
```
```suggestion
        /// LHS must evaluate to a DateTime
```

sharkdp commented on numbat/src/typed_ast.rs:
```diff
@@ -109,6 +122,13 @@ impl PrettyPrint for &Vec<StringPart> {
+pub enum DateOperationResult {
+    /// An operation with datetimes that produces something a value in Seconds
+    Seconds,
+    Other,
```
What is the other possible result? A DateTime, right?

sharkdp commented on numbat/src/vm.rs:
```diff
@@ -743,6 +834,7 @@ impl Vm {
                             Value::Quantity(q) => q.to_string(),
                             Value::Boolean(b) => b.to_string(),
                             Value::String(s) => s,
+                            Value::DateTime(..) => "DateTimeTodo_1".to_string(),
```
This should be implemented for use cases like:
```
print("Current time: {now()}")
```

eminence replied:
Ahh, I totally failed to notice that string interpolation uses this different codepath.  This should now be fixed

sharkdp commented on numbat/src/bytecode_interpreter.rs:
```diff
@@ -113,6 +119,28 @@ impl BytecodeInterpreter {
+                self.compile_expression(lhs)?;
+                self.compile_expression(rhs)?;
+
+                let op = if *result_type == DateOperationResult::Seconds {
```
Could we figure this out from `_type`? This way, we might be able to get rid of `result_type`?

eminence replied:
Oh, good point, yes I think we can.  I will try this.  If so, that would be much cleaner

sharkdp commented on numbat/src/typechecker.rs:
```diff
@@ -487,134 +497,206 @@ impl TypeChecker {
+                    let rhs_is_datetime = rhs_checked.get_type() == Type::DateTime;
+
+                    if *op == BinaryOperator::ConvertTo
+                        && matches!(rhs_checked, typed_ast::Expression::String(..))
```
Is there any reason why you implemented it like this, compared to
```suggestion
                        && rhs_checked.get_type() == Type::String
```
which would allow us to do things like:
```py
let tz_germany = "Europe/Berlin"
now() -> tz_germany
```

eminence replied:
No good reason that I can recall 😃

mhmdanas commented on numbat-cli/src/completer.rs:
```diff
@@ -73,6 +74,45 @@ impl Completer for NumbatCompleter {
+            .or_else(|| line.find("→"))
+            .or_else(|| line.find("➞"))
+            .or_else(|| line.find(" to "))
+            .map_or(None, |convert_pos| {
```
```suggestion
            .and_then(|convert_pos| {
```

should be the same, I think.

mhmdanas commented on numbat/src/ffi.rs:
```diff
@@ -740,3 +785,59 @@ fn chr(args: &[Value]) -> Result<Value> {
 }
+
+fn now(args: &[Value]) -> Result<Value> {
+    assert!(args.len() == 0);
```
```suggestion
    assert!(args.is_empty());
```

mhmdanas commented on numbat/src/ffi.rs:
```diff
@@ -740,3 +785,59 @@ fn chr(args: &[Value]) -> Result<Value> {
+    // Try to parse as rfc3339 and if that fails then as rfc2822
+    let output = chrono::DateTime::parse_from_rfc3339(input)
+        .or_else(|_| chrono::DateTime::parse_from_rfc2822(input))
+        .map_err(|e| RuntimeError::DateParsingError(e))?;
```
```suggestion
        .map_err(RuntimeError::DateParsingError)?;
```

mhmdanas commented on numbat/src/vm.rs:
```diff
@@ -743,6 +831,14 @@ impl Vm {
+                                        dt.naive_utc(),
+                                        offset,
+                                    );
+                                format!("{}", l.to_rfc2822())
```
```suggestion
                                l.to_rfc2822()
```

mhmdanas reviewed (changes requested):
Thank you! This looks cool, just a few minor nitpicks.

sharkdp commented on numbat/src/vm.rs:
```diff
@@ -615,6 +649,60 @@ impl Vm {
+                    let rhs = self.pop_datetime();
+                    let lhs = self.pop_datetime();
+
+                    let duration = lhs - rhs;
+                    // TODO error handling
+                    let duration = duration.num_nanoseconds().unwrap() as f64 / 1_000_000_000f64;
```
`num_nanoseconds` returns its result as an `i64`. 2^63 nanoseconds are "only" 292 years, so currently we'll get a panic here if we attempt to subtract two datetimes that are further apart than that:
```
>>> now() - parse_datetime("1731-01-01T12:00:00Z")
thread 'main' panicked at numbat/src/vm.rs:685:63:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

eminence replied:
Ok good news, if we upgrade chrono, I think we can do this new function: 

```suggestion
                    let duration = lhs - rhs;
                    let duration = duration.subsec_nanos() as f64 / 1_000_000_000f64 + duration.num_seconds() as f64;
```

sharkdp reviewed (approved):
Thank you so much!
